### ISSUE TITLE
[QUESTION] How to Properly Capture grad_output in RowParallelLinear and ColumnParallelLinear?

### ISSUE THREAD

**Issue by UnbearableFate** [NVIDIA/Megatron-LM#1629]
Hi all,

I’m working on implementing a K-FAC-style preconditioner in Megatron-LM, and I need to capture the output gradients (grad_output) from layers like RowParallelLinear and ColumnParallelLinear during backpropagation.

I’ve tried using module.register_full_backward_hook(...) to intercept grad_output for each module like the following: 
`module.register_full_backward_hook(self._save_grad_output)`

```
class Kfac_preconditioner 

    def __init__ (self):
        # do something
        module.register_full_backward_hook(self._save_grad_output)


    @torch.no_grad()
    def _save_grad_output(
            self,
            module: torch.nn.Module,
            grad_input: tuple[torch.Tensor, ...] | torch.Tensor,
            grad_output: tuple[torch.Tensor, ...] | torch.Tensor,
    ) -> None:
        """Hook for saving the gradient w.r.t. output in the backward pass."""
        if not module.training:
            return
        if self.steps % self.factor_update_steps == 0:
            name, layer = self._layers[module]
            if isinstance(grad_output, torch.Tensor):
                grad_output = (grad_output,)
            layer.save_layer_grad_output(grad_output)
            layer.update_g_factor(alpha=self.factor_decay)
```
      

 but this doesn’t seem to work reliably—for most time the gradient tensors contain only NaN/inf values.

So I have a few questions:
	1.	What is the correct way to access grad_output for RowParallelLinear and ColumnParallelLinear layers?
	2.	Should I register a backward hook on the forward output tensor instead (via .register_hook())?

Any advice, references, or code snippets would be greatly appreciated!

Thanks in advance.

PS: I’m currently only working with GPT-style models using default settings, and tensor parallelism only (no pipeline parallel or sequence parallel).
However, if you could also point out how to properly handle this when pipeline parallelism or sequence parallelism is enabled, I’d greatly appreciate it.

**Comment by UnbearableFate** [comment]
The use of `register_full_backward_hook` was not actually the problem. The real cause of the covariance matrix of the gradients being filled with NaN and Inf values was the default use of FP16 training, which led to numerical overflow.
